### PR TITLE
Fixing model checkpoints to be robust to mutli -> single GPU usage

### PR DESCRIPTION
**Pull request by pruksmhc**
This is a fix to https://github.com/nyu-mll/jiant/issues/1087. I decided to make the change in the model loading portion because making the change in model saving as suggested in the https://github.com/nyu-mll/jiant/issues/1087 will fix multi -> single GPU model loading, but will break multi -> multi GPU model loading (if we want to reload a checkpoint that was trained in multi-GPU on a multi-GPU machine). 
Additionally, I also did some light cleanup of model loading in trainer to not be redundant, and also deleted an unused parameter. 

**Tests**
Multi -> Single GPU: I tested by training a roberta-large model on SST on multi-GPU, and then loading that checkpoint in a single-GPU for further training. 
Multi -> Multi GPU: This is implicitly already done in jiant, specifically we load the best checkpoint before doing evaluation, so this was tested when I trained the roberta-large SST model the first time on multi-GPU.

### CODE REVIEW

File: jiant/utils/utils.py [hunk 1]
```diff
@@ -323,22 +325,48 @@ def select_task_specific_args(exp_args, diff_args):
...
+    final_model_state = collections.OrderedDict()
+
+    def get_key(name):
+        key = name
+        if isinstance(model, nn.DataParallel):
+            if "module" not in key:
+                key = "module.%s" % key
+        else:
+            if "module" in key:
+                key = key.replace("module.", "")
```
Comment (zphang) [inline jiant/utils/utils.py:346]: Make this check explicitly for prefix with `.startswith`, and drop the first n characters (in case module appears somewhere else in the parameter name).

File: jiant/utils/utils.py [hunk 2]
```diff
@@ -323,22 +325,48 @@ def select_task_specific_args(exp_args, diff_args):
...
+        return key
+
+    for name, weights in model_state.items():
+        key = get_key(name)
+        final_model_state[key] = model_state[name]
```
Comment (zphang) [inline jiant/utils/utils.py:351]: ` = weights`

File: jiant/trainer.py [hunk 3]
```diff
@@ -1217,15 +1217,7 @@ def _restore_checkpoint(self, phase, tasks=None):
...
-        model_state = torch.load(model_path)
-
-        for name, param in self._model.named_parameters():
-            if param.requires_grad and name not in model_state:
-                log.error("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!")
-                log.error("Parameter missing from checkpoint: " + name)
-                log.error("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!")
-
-        self._model.load_state_dict(model_state, strict=False)
+        load_model_state(self._model, model_path)
```
Comment (zphang) [inline jiant/trainer.py:1220]: Is this warning being disabled?

Comment (pruksmhc) [reply jiant/trainer.py:1220]: No it's not. This warning is also inside load_model_state function

**Comment by pyeres** [comment]
Hi @zphang & @HaokunLiu — are either of you available to provide the substantial review for this PR? The core concerns seem to be 1) whether this addresses issue #1087, and 2) whether these changes introduce new risks/regressions.

File: jiant/utils/utils.py [hunk 4]
```diff
@@ -323,22 +325,49 @@ def select_task_specific_args(exp_args, diff_args):
...
+        multi-GPU and single-GPU use cases
+
+    Parameters
+    ----------
+    model: The model object to populate with loaded parameters.
+    model_state: collections.OrderdDict of model state
+    """
+    final_model_state = collections.OrderedDict()
+
+    def get_key(name):
```
Comment (sleepinyourhat) [inline jiant/utils/utils.py:339]: Add a comment explaining why we need this logic.

File: jiant/trainer.py [hunk 5]
```diff
@@ -1217,15 +1217,7 @@ def _restore_checkpoint(self, phase, tasks=None):
...
 
-        model_state = torch.load(model_path)
-
-        for name, param in self._model.named_parameters():
-            if param.requires_grad and name not in model_state:
-                log.error("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!")
-                log.error("Parameter missing from checkpoint: " + name)
-                log.error("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!")
-
-        self._model.load_state_dict(model_state, strict=False)
```
Comment (sleepinyourhat) [inline jiant/trainer.py:1228]: You're no longer setting strict=False here. It's debatable whether that's the ideal behavior here, but it was intentional, and I believe it has had some real experimental uses. Why the change?

**Comment by jeswan** [comment]
Are these changes still necessary? Planning to close all PRs to move jiant2 to this repo in the near future.
